From c26f2da15edc27bbdb13bf5e3f7d9deb8504c317 Mon Sep 17 00:00:00 2001
From: Michal Simek <michal.simek@xilinx.com>
Date: Wed, 17 Jun 2020 15:27:17 +0200
Subject: [PATCH 1390/1852] clk: zynqmp: Add missing checking of eemi_ops

commit 79785f65d157e110e033284f3eaec2dce84b7e59 from
https://github.com/Xilinx/linux-xlnx.git

Check if eemi_ops are properly initialized. If not error out.

The issue is reported by smatch like
  CHECK   drivers/clk/zynqmp/clk-mux-zynqmp.c
drivers/clk/zynqmp/clk-mux-zynqmp.c:52 zynqmp_clk_mux_get_parent()
 error: 'eemi_ops' dereferencing possible ERR_PTR()
drivers/clk/zynqmp/clk-mux-zynqmp.c:76 zynqmp_clk_mux_set_parent()
 error: 'eemi_ops' dereferencing possible ERR_PTR()

Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/clk/zynqmp/clk-mux-zynqmp.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/zynqmp/clk-mux-zynqmp.c b/drivers/clk/zynqmp/clk-mux-zynqmp.c
index 9f43c25b5a8a..e1493c75f3a8 100644
--- a/drivers/clk/zynqmp/clk-mux-zynqmp.c
+++ b/drivers/clk/zynqmp/clk-mux-zynqmp.c
@@ -49,8 +49,10 @@ static u8 zynqmp_clk_mux_get_parent(struct clk_hw *hw)
 	int ret;
 	const struct zynqmp_eemi_ops *eemi_ops = zynqmp_pm_get_eemi_ops();
 
-	ret = eemi_ops->clock_getparent(clk_id, &val);
+	if (IS_ERR(eemi_ops))
+		return PTR_ERR(eemi_ops);
 
+	ret = eemi_ops->clock_getparent(clk_id, &val);
 	if (ret)
 		pr_warn_once("%s() getparent failed for clock: %s, ret = %d\n",
 			     __func__, clk_name, ret);
@@ -73,8 +75,10 @@ static int zynqmp_clk_mux_set_parent(struct clk_hw *hw, u8 index)
 	int ret;
 	const struct zynqmp_eemi_ops *eemi_ops = zynqmp_pm_get_eemi_ops();
 
-	ret = eemi_ops->clock_setparent(clk_id, index);
+	if (IS_ERR(eemi_ops))
+		return PTR_ERR(eemi_ops);
 
+	ret = eemi_ops->clock_setparent(clk_id, index);
 	if (ret)
 		pr_warn_once("%s() set parent failed for clock: %s, ret = %d\n",
 			     __func__, clk_name, ret);
-- 
2.31.1

